src/app/components/import/ImportingModalContent.js

```diff
@@ -0,0 +1,150 @@
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import { c } from 'ttag';
+import { Alert, useApi } from 'react-components';
+
+import DynamicProgress from '../DynamicProgress';
+
+import { addContacts } from 'proton-shared/lib/api/contacts';
+import { extractVcards, parse as parseVcard } from '../../helpers/vcard';
+import { prepareContact } from '../../helpers/encrypt';
+import { percentageProgress } from '../../helpers/progress';
+import { OVERWRITE, CATEGORIES, SUCCESS_IMPORT_CODE } from '../../constants';
+import { CONTACT_CARD_TYPE } from 'proton-shared/lib/constants';
+
+const { OVERWRITE_CONTACT } = OVERWRITE;
+const { IGNORE, INCLUDE } = CATEGORIES;
+const { CLEAR_TEXT } = CONTACT_CARD_TYPE;
+
+const [PARSING, ENCRYPTING, IMPORTING, FINISHED] = [1, 2, 3, 4];
+
+const ImportingModalContent = ({
+    fileType,
+    file,
+    vcardContacts,
+    onSetVcardContacts,
+    loadingKeys,
+    privateKey,
+    onEncryptingDone
+}) => {
+    const api = useApi();
+
+    const [step, setStep] = useState(PARSING);
+    const [track, setTrack] = useState({
+        total: 0,
+        encrypted: [],
+        failedOnParse: 0,
+        failedOnEncrypt: [],
+        imported: 0
+    });
+
+    const parseFileIfNeeded = async () => {
+        if (fileType === 'text/vcard') {
+            const vcards = extractVcards(file);
+            setTrack((track) => ({ ...track, total: vcards.length }));
+            const parsedVcards = [];
+            for (const vcard of vcards) {
+                try {
+                    parsedVcards.push(parseVcard(vcard));
+                } catch {
+                    setTrack((track) => ({ ...track, failedOnParse: track.failedOnParse + 1 }));
+                }
+            }
+            onSetVcardContacts(parsedVcards);
+        } else {
+            // in the case of a csv file, the parsing has occurred in the previous step
+            setTrack((track) => ({ ...track, total: vcardContacts.length }));
+        }
+        setStep(ENCRYPTING);
+    };
+
+    const encryptContacts = async () => {
+        const publicKey = privateKey.toPublic();
+        for (const vcardContact of vcardContacts) {
+            try {
+                const contactEncrypted = await prepareContact(vcardContact, privateKey, publicKey);
+                setTrack((track) => ({ ...track, encrypted: [...track.encrypted, contactEncrypted] }));
+            } catch (error) {
+                setTrack((track) => ({ ...track, failedOnEncrypt: [...track.failed, vcardContact] }));
+            }
+        }
+        onEncryptingDone();
+        setStep(IMPORTING);
+    };
+
+    const importContacts = async (contacts) => {
```

Comment:
There should be a way to extract this contact logic outside of the component as hooks or even pure functions. It would be easier to test and to maintain.

Reply:
It's true there's quite a bit of code inside the component, but because it's very specific to this component and it will not be re-used outside, we thought it's ok to leave it inside.